### DISCUSSION TITLE
Examples?

### DISCUSSION THREAD

**Discussion by jfmatth** [kubero-dev/kubero#103] [Ideas/Features]
This project may be what I've been looking for, but ... there don't seem to be any examples for the various buildpacks.

Can we get a Python one posted?  I'm happy to help build a repo for this, but honestly don't know what is required for Kubero to work ?

Any help is appreciated :)

**Comment by mms-gianni** [comment]
I am more than happy to help. 

There are 2 strategies to deploy your code: git (with the build packs) and docker (prebuilt images)

Git strategy:
Buildpacks may be a bit misleading since it has nothing in common with buildpacks.io . Buildpacks are just a set of 3 containers. Those containers use the official images and don't need any maintenance. 

1. fetch : you regularly don't want to replace this one, since it just fetches your code with a git checkout
2. build : this container does your "pip install" 
3. run : this container mounts the volume with the build application in step 2 in read-only mode and is your remaining container. 

Here is a plain Python example with docker-compose to explain those 3 steps. Kubero uses the same technique with deployments and pods. 
https://github.com/kubero-dev/buildpacks/blob/main/packs/python/docker-compose.yaml 

This git strategy is very fast. But it does not scale very well and becomes more inefficient with every additional container. 

Docker(prebuilt) strategy;
This strategy scales better and is more the traditional way to deploy a code. In this strategy, you don't have these 3 steps. You will need to build your image in your classic CI/CD process (Github Actions for example) . 
(build Dockerimage ->push to registry -> pull in Kubernetes)

**Reply by jfmatth** [reply]
@mms-gianni thank you so much for the quick response.

I'll play with these examples and see if i have more questions :)

I've been longing for a git push system ever since they retired Openshift 2 😄 

J

**Reply by jfmatth** [reply]
@mms-gianni 

First off, incredible project.  Seems like it takes some very simple pieces and stiches them all together to make a great PAAS.

However, i'm still unable to get it working ☹️ 

I forked the python repo it got from this thread, https://github.com/jfmatth/kubero-python and adjusted the run portion to account for main.py, but I get init errors when it deploys the app.

I tried looking for the logs but was unable to find anything, and all Kubero told me was error deploying pod.

I'm happy to take what I learn and build an example Python / Kubernetes repo to help others, but need some help here please 😄 

thanks,

**Reply by mms-gianni** [reply]
Sure. 

I have to admit, I'm currently mostly coding with NodeJS. But I deployed my example without any issues. 

Are you sure you've pushed your changes to your repository? I could not find main.py 

These are my logs: 

```
~ [ kubectl logs pod/python-kuberoapp-web-7b66b9c6c4-5qxs8 -c kuberoapp-fetcher                                                                                                                 ] 5:26 pm
========== whipe the app dir
========== Clone Repository from git@github.com:kubero-dev/template-python.git
Cloning into '.'...
Failed to add the host to the list of known hosts (/root/.ssh/known_hosts).
Already on 'main'
Your branch is up to date with 'origin/main'.
========== write startupscripts based on Procfile
init-build.sh
init-web.sh
init-worker.sh



~ [ kubectl logs pod/python-kuberoapp-web-7b66b9c6c4-5qxs8 -c kuberoapp-builder                                                                                                                 ] 5:27 pm

[notice] A new release of pip available: 22.3.1 -> 23.0
[notice] To update, run: pip install --upgrade pip
```

**Reply by jfmatth** [reply]
Tried again this AM, with more information

Built a pipeline called home, with an app pythontest.  it fails on 'init'ing the container, see below

```jfmatth@jfmatth-gaming:~$ k get all -n home-test
NAME                                            READY   STATUS                  RESTARTS      AGE
pod/pythontest-kuberoapp-web-5c749f8cb4-zzvzx   0/1     Init:CrashLoopBackOff   4 (86s ago)   3m32s

NAME                           TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
service/pythontest-kuberoapp   ClusterIP   10.43.240.115   <none>        80/TCP    3m32s

NAME                                          READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/pythontest-kuberoapp-web      0/1     1            0           3m32s
deployment.apps/pythontest-kuberoapp-worker   0/0     0            0           3m32s

NAME                                                     DESIRED   CURRENT   READY   AGE
replicaset.apps/pythontest-kuberoapp-web-5c749f8cb4      1         1         0       3m32s
replicaset.apps/pythontest-kuberoapp-worker-58f4c58764   0         0         0       3m32s
jfmatth@jfmatth-gaming:~$ k logs -n home-test pod/pythontest-kuberoapp-web-5c749f8cb4-zzvzx
Defaulted container "kuberoapp-web" out of: kuberoapp-web, kuberoapp-fetcher (init), kuberoapp-builder (init)
Error from server (BadRequest): container "kuberoapp-web" in pod "pythontest-kuberoapp-web-5c749f8cb4-zzvzx" is waiting to start: PodInitializing
```

Not sure what I'm doing wrong, seems simple enough

**Reply by jfmatth** [reply]
@mms-gianni 

Thanks for the quick reply.  

Yes, my repo didn't reflect the main.py, thought it did.

I'm still unable to get it just past the init phase, which seems odd.  Pulling the fetch container failed on the command line too ☹️ 
```
docker pull ghcr.io/kubero-dev/buildpacks/fetch:sha256-5625767a9cf9cdb972826b1713f83b4a50f441c8b88ce5eb6f923d81121daee1.sig
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Trying to pull ghcr.io/kubero-dev/buildpacks/fetch:sha256-5625767a9cf9cdb972826b1713f83b4a50f441c8b88ce5eb6f923d81121daee1.sig...
Getting image source signatures
Copying blob 83b63d88ade9 done
Error: writing blob: adding layer with blob "sha256:83b63d88ade91a8cb899fda9a3b0c2a1a9a9a361f663585905458b7d0528edc1": Error processing tar file(exit status 1): unexpected EOF
```

**Reply by mms-gianni** [reply]
To view te logs of the init container you need to add the "-c" flag: 
`kubectl logs pod/python-kuberoapp-web-7b66b9c6c4-5qxs8 -c kuberoapp-fetcher`
and for the builder:
`kubectl logs pod/python-kuberoapp-web-7b66b9c6c4-5qxs8 -c kuberoapp-builder`

Not able to pull is strange tough ... I need to investigate this.

**Reply by jfmatth** [reply]
updated the repo online to main.py, still having init errors.  I don't know why, what am I missing?  

```
 k logs -n home-test python-kuberoapp-web-688cccf746-b9wjp
Defaulted container "kuberoapp-web" out of: kuberoapp-web, kuberoapp-fetcher (init), kuberoapp-builder (init)
Error from server (BadRequest): container "kuberoapp-web" in pod "python-kuberoapp-web-688cccf746-b9wjp" is waiting to start: PodInitializing
jfmatth@jfmatth-gaming:~/development/kubero/kubero-python$ k get pods -n home-test -w
NAME                                    READY   STATUS       RESTARTS      AGE
python-kuberoapp-web-688cccf746-b9wjp   0/1     Init:Error   2 (20s ago)   33s
python-kuberoapp-web-688cccf746-b9wjp   0/1     Init:CrashLoopBackOff   2 (13s ago)   44s
python-kuberoapp-web-688cccf746-b9wjp   0/1     Init:1/2                3 (26s ago)   57s
python-kuberoapp-web-688cccf746-b9wjp   0/1     Init:Error              3 (29s ago)   60s
python-kuberoapp-web-688cccf746-b9wjp   0/1     Init:CrashLoopBackOff   3 (13s ago)   73s
```

**Reply by jfmatth** [reply]
You're logs gave me a clue, looks like i can't read my repo?

```
jfmatth@jfmatth-gaming:~/development/kubero/kubero-python$ k logs -n home-test python-kuberoapp-web-688cccf746-b9wjp -c kuberoapp-fetcher
========== whipe the app dir
========== Clone Repository from git@github.com:jfmatth/kubero-python.git
Cloning into '.'...
Failed to add the host to the list of known hosts (/root/.ssh/known_hosts).
Load key "/root/.ssh/deploykey": invalid format
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
========== write startupscripts based on Procfile
init-build.sh
init-web.sh
init-worker.sh
```

is there a way not to use ssh into GH?

**Comment by mms-gianni** [comment]
Back to the git strategy: 
Kubero has a set of preconfigured "buildpacks" here: 
https://github.com/kubero-dev/kubero/blob/main/config.yaml#L36-L48 

You can extend or replace them in your running kubero config (modify configmap, restart kubero UI container) 

And here is the switch to chose between those 2 strategies: 
![image](https://user-images.githubusercontent.com/2052196/218873893-6b34d5e8-fe9f-4056-8089-a228a4882fde.png)

**Reply by jfmatth** [reply]
this part makes sense, and I just need to know what files would affect the build process.

**Reply by mms-gianni** [reply]
Same as on Heroku: the Procfile 

This is my Python example: 
https://github.com/kubero-dev/template-python/blob/main/Procfile 

And here how it is processed: 
https://github.com/kubero-dev/buildpacks/blob/main/packs/fetch/build/init.sh#L20

**Reply by jfmatth** [reply]
so far I've determined that since Kubero uses SSH into GH, you'll need a keypair there, but don't see where to add that, hmm.  Checking docs now.

i'm also going to update the config to see if i can force the repo to be https for now :)
